soc/arm/st_stm32/stm32l0/power.c

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2021 Fabio Baltieri
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <pm/pm.h>
+#include <soc.h>
+#include <init.h>
+
+#include <stm32l0xx_ll_utils.h>
+#include <stm32l0xx_ll_bus.h>
+#include <stm32l0xx_ll_cortex.h>
+#include <stm32l0xx_ll_pwr.h>
+#include <stm32l0xx_ll_rcc.h>
+#include <stm32l0xx_ll_system.h>
+#include <clock_control/clock_stm32_ll_common.h>
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+
+/* Select MSI as wake-up system clock if configured, HSI otherwise */
+#if STM32_SYSCLK_SRC_MSI
+#define RCC_STOP_WAKEUPCLOCK_SELECTED LL_RCC_STOP_WAKEUPCLOCK_MSI
+#else
+#define RCC_STOP_WAKEUPCLOCK_SELECTED LL_RCC_STOP_WAKEUPCLOCK_HSI
+#endif
+
+/* Invoke Low Power/System Off specific Tasks */
+void pm_power_state_set(struct pm_state_info info)
+{
+	switch (info.state) {
+	case PM_STATE_SUSPEND_TO_IDLE:
+		LL_RCC_SetClkAfterWakeFromStop(RCC_STOP_WAKEUPCLOCK_SELECTED);
+		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP);
```

Comment:
I see to interesting commands in the stm32cube example that controls some regulators. Are they influencing the power consumption there also ?

```
  /* Enable ultra low power mode */
  LL_PWR_EnableUltraLowPower();

  /** Set the regulator to low power before setting MODE_STOP.   
    * If the regulator remains in "main mode",   
    * it consumes more power without providing any additional feature. */    
  LL_PWR_SetRegulModeLP(LL_PWR_REGU_LPMODES_LOW_POWER);

```

Reply:
This makes huge difference. In my test it's a drop in sleep current from 16.9 uA down to 1.4 uA.

From my understanding this is a one-time setup thing so it should suffice to do it in `stm32_power_init`. Maybe it should even be up to the application or board to perform this setup in an init routine?

Reply:
Cool! Yeah works great... if it's there in the cube code I don't see why not put it in here as suggested, I'll add it. I think it's good to have it in this code path, juts in case the the PM code gets more complex and this bit ends up being changed somewhere else... keep the prep code together may avoid regressions.